### ISSUE TITLE
[Impeller] Svg with linear gradient rendering incorrectly.

### ISSUE THREAD

**Issue by eldrigo** [flutter/flutter#140970]
### Steps to reproduce

I have an svg with shadows which is attached below and it does not render correctly on IOS
but the SVG is rendered correctly when running without **impeller** 

use the attached SVG with 
`flutter_svg: ^2.0.9` 

<img src="https://github.com/flutter/flutter/assets/33096241/1f90144c-c748-402d-b751-6af175d29883.svg"  height=200 widht=200>

| Expected results | Actual results |
| -- | -- |
| This should be the expected output (running with skia) | This is the result with impeller |
| <img src="https://github.com/flutter/flutter/assets/33096241/6307de79-8b8a-406b-a19f-137dc688f008.png" height=700> | <img src="https://github.com/flutter/flutter/assets/33096241/b17acd23-5fae-4ebe-91a5-bc0b3feaec3d" height=700> |

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_svg/svg.dart';

void main() {
  runApp(const SvgTest());
}

class SvgTest extends StatelessWidget {
  const SvgTest({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
            child: SvgPicture.asset(
          'assets/test.svg',
          height: 300,
          width: 300,
        )),
      ),
    );
  }
}
```
### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.16.5, on macOS 14.1.1 23B81 darwin-arm64, locale en-GB)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.1)
[✓] VS Code (version 1.85.1)
[✓] Connected device (3 available)
```

</details>

**Comment by danagbemava-nc** [comment]
Reproducible using the code sample and SVG provided above. 

<details>
<summary>flutter doctor -v</summary>

```
[!] Flutter (Channel stable, 3.16.5, on macOS 14.1.2 23B92 darwin-arm64, locale en-GB)
    • Flutter version 3.16.5 on channel stable at /Users/nexus/dev/sdks/flutter
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter_rc/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutter. Consider adding /Users/nexus/dev/sdks/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter_rc/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutter. Consider adding /Users/nexus/dev/sdks/flutter/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 78666c8dc5 (2 weeks ago), 2023-12-19 16:14:14 -0800
    • Engine revision 3f3e560236
    • Dart version 3.2.3
    • DevTools version 2.28.4
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0-rc1)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0-rc1
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0)
    • Xcode at /Applications/Xcode-15.0.0-Release.Candidate.app/Contents/Developer
    • Build 15A240d
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.5)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 76.3.4
    • Dart plugin version 232.10072.19

[✓] VS Code (version 1.85.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.80.0

[✓] Connected device (5 available)
    • Pixel 7 (mobile)     • 28291FDH2001SA            • android-arm64  • Android 14 (API 34)
    • Nexus (mobile)       • 00008020-001875E83A38002E • ios            • iOS 17.2.1 21C66
    • Dean’s iPad (mobile) • 00008103-000825C811E3401E • ios            • iOS 17.2 21C62
    • macOS (desktop)      • macos                     • darwin-arm64   • macOS 14.1.2 23B92 darwin-arm64
    • Chrome (web)         • chrome                    • web-javascript • Google Chrome 120.0.6099.199

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
 
```
[!] Flutter (Channel master, 3.18.0-19.0.pre.76, on macOS 14.1.2 23B92 darwin-arm64, locale en-GB)
    • Flutter version 3.18.0-19.0.pre.76 on channel master at /Users/nexus/dev/sdks/flutters
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter_rc/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter_rc/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 89809bef71 (3 hours ago), 2024-01-05 00:15:24 -0500
    • Engine revision 7e63486409
    • Dart version 3.3.0 (build 3.3.0-279.0.dev)
    • DevTools version 2.31.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0-rc1)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0-rc1
    • Java binary at: /Users/nexus/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0)
    • Xcode at /Applications/Xcode-15.0.0-Release.Candidate.app/Contents/Developer
    • Build 15A240d
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Users/nexus/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2.5)
    • IntelliJ at /Users/nexus/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 76.3.4
    • Dart plugin version 232.10072.19

[✓] VS Code (version 1.85.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.80.0

[✓] Connected device (5 available)
    • Pixel 7 (mobile)     • 28291FDH2001SA            • android-arm64  • Android 14 (API 34)
    • Nexus (mobile)       • 00008020-001875E83A38002E • ios            • iOS 17.2.1 21C66
    • Dean’s iPad (mobile) • 00008103-000825C811E3401E • ios            • iOS 17.2 21C62
    • macOS (desktop)      • macos                     • darwin-arm64   • macOS 14.1.2 23B92 darwin-arm64
    • Chrome (web)         • chrome                    • web-javascript • Google Chrome 120.0.6099.199

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>

**Comment by dnfield** [comment]
SVG is

```svg
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 56 56">
  <defs>
    <linearGradient id="a" x1="-3.86" x2="43.76" y1="-3.79" y2="43.09" data-name="Unbenannter Verlauf 3"
      gradientUnits="userSpaceOnUse">
      <stop offset=".13" stop-color="#ffb800" stop-opacity="0" />
      <stop offset=".99" stop-color="#ffb800" />
    </linearGradient>
    <linearGradient id="b" x1="44.59" x2="43.6" y1="28.47" y2="22.56" data-name="Unbenannter Verlauf 6"
      gradientUnits="userSpaceOnUse">
      <stop offset=".13" stop-color="#9cdaf5" stop-opacity="0" />
      <stop offset=".99" stop-color="#44baec" />
    </linearGradient>
  </defs>
  <g style="isolation:isolate">
    <g data-name="Ebene 4">
      <circle cx="28" cy="28" r="27.43" fill="#ffde52" />
      <circle cx="28.98" cy="28.55" r="23.8" fill="url(#a)" opacity=".5" style="mix-blend-mode:multiply" />
    </g>
    <path fill="none" stroke="#2e2d2c" stroke-linecap="round" stroke-linejoin="round" stroke-width="2.5"
      d="M41.63 32.87a2.48 2.48 0 1 1-4.96 0m-17.39 0a2.48 2.48 0 1 1-4.96 0m8.63 10.87c1.25-.51 3.16-1.11 5.54-1.24 2.85-.15 5.14.44 6.52.91"
      data-name="1" />
    <g data-name="5">
      <path fill="#44baec"
        d="M39.54 24.22c0-2.39 4.33-7.7 4.33-7.7s4.33 5.31 4.33 7.7-1.94 4.33-4.33 4.33-4.33-1.94-4.33-4.33" />
      <circle cx="43.87" cy="24.17" r="2.83" fill="url(#b)" />
    </g>
  </g>
</svg>
```

**Comment by jonahwilliams** [comment]
I took a look at the frame caputre and I suspect that this is actually a bug in the linear gradient shader. The offscreen of the gradient is already quite dark:

![image](https://github.com/flutter/flutter/assets/8975114/4d9af444-f15e-4a02-afca-4c2ad1d2678f)

Since this gradient should be going from fully opaque to slightly grey, it makes me think we're either computing the color incorrectly, or the offsets incorrectly.

**Comment by jason-simmons** [comment]
I noticed that this renders as intended if I patch `framebuffer_blend.frag` to set `src = IPHalfUnpremultiply(src)`

**Comment by bdero** [comment]
Isolated this to advanced blends in general (repros for both fallback advanced blends and the SSBO path).

Minimal svg with just a solid color circle:
```svg
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 56 56">
  <defs>
  </defs>
  <g style="isolation:isolate">
    <circle cx="28.98" cy="28.55" r="23.8" fill="#ffde5222" opacity="1" style="mix-blend-mode:multiply" />
  </g>
</svg>
```

Impeller (left), Skia (right):
![image](https://github.com/flutter/flutter/assets/919017/3c8664c5-4431-4605-ae49-3fb7a1455585)

There's definitely a double-premultiply going on here as @jason-simmons's findings suggest, but I haven't worked out the proper solution quite yet.
All source images fed to filters are expected to be premultipled in Impeller (since ui.Images are premultiplied while decoding, and all rendered content in Impeller is premultiplied. I think my [earlier patch](https://github.com/flutter/engine/commit/90d48e1e4a7297b03631cae7a7b06f469ce684ce) is making incorrect assumptions about the necessary ceremony around the blur functions.

In other news, I added another case to the blend test and placed an identical checkerboard behind the images as displayed the Flutter BlendMode docs, and I spotted an alpha management problem present in all advanced blends. I don't know if it's related yet.

Impeller (left), Skia (right):
![image](https://github.com/flutter/flutter/assets/919017/5d951b28-70f3-4c88-b8ff-900b009b7841)

Notice the transparent output in Impeller.
